File: code-examples/scala/transactions/txn-convenient.scala
import org.mongodb.scala._
import org.mongodb.scala.bson.collection.immutable.Document
import org.mongodb.scala.model.Filters._
import org.mongodb.scala.model.Updates._
import scala.concurrent.Await
import scala.concurrent.duration._
import scala.concurrent.ExecutionContext.Implicits.global

val mongoClient: MongoClient = MongoClient("<connection string uri>")
val database: MongoDatabase = mongoClient.getDatabase("testdb")

val inventoryCollection: MongoCollection[Document] = database.getCollection("inventory")
val ordersCollection: MongoCollection[Document] = database.getCollection("orders")

val inventoryData: Seq[Document] = Seq(
  Document("item" -> "sunblock", "qty" -> 85, "price" -> 6.0),
  Document("item" -> "beach chair", "qty" -> 30, "price" -> 25.0)
)
inventoryCollection.insertMany(inventoryData)

val order: Seq[Document] = Seq(
  Document("item" -> "sunblock", "qty" -> 3),
  Document("item" -> "beach chair", "qty" -> 1)
)

val session: ClientSession = mongoClient.startSession()

try {
  session.startTransaction()

  var total: Double = 0

  order.foreach { item =>
    val updateResult = Await.result(inventoryCollection.findOneAndUpdate(
      session,
      and(equal("item", item.getString("item")), gte("qty", item.getInteger("qty"))),
      inc("qty", -item.getInteger("qty"))
    ).head(), 10.seconds)

    if (updateResult == null) {
      session.abortTransaction()
      println("Item not found or insufficient quantity.")
      return
    }

    val subTotal: Double = item.getInteger("qty") * updateResult.getDouble("price")
    total = total + subTotal
  }

  val receipt = Document(
    "date" -> java.time.LocalDateTime.now(),
    "items" -> order,
    "total" -> total
  )

  Await.result(ordersCollection.insertOne(session, receipt).head(), 10.seconds)

  println(
    "Order successfully completed and recorded!\nReceipt:\n" +
      receipt.toJson()
  )

  session.commitTransaction()
} catch {
  case e: Exception =>
    println("Error processing transaction", e)
    session.abortTransaction()
} finally {
  session.close()
  mongoClient.close()
}

File: code-examples/php/transactions/txn-convenient.php
<?php
require 'vendor/autoload.php';

$uri = '<connection string uri>';
$client = new MongoDB\Client($uri);
$db = $client->testdb;
$inventory = $db->inventory;

// Insert the inventory
$inventory->insertMany([
  ['item' => 'sunblock', 'qty' => 85, 'price' => 6.0],
  ['item' => 'beach chair', 'qty' => 30, 'price' => 25.0]
]);

$orders = $db->orders;

$order1 = [
  ['item' => 'sunblock', 'qty' => 3],
  ['item' => 'beach chair', 'qty' => 1]
];

$order2 = [
  ['item' => 'volleyball', 'qty' => 1]
];

$order = $order1;

$session = $client->startSession();

try {
  $session->startTransaction();

  $total = 0;
  foreach ($order as $item) {
    $itemForUpdate = $inventory->findOne([
      'item' => $item['item'],
      'qty' => ['$gte' => $item['qty']]
    ]);

    if ($itemForUpdate === null) {
      $session->abortTransaction();
      $session->endSession();
      echo "Item not found or insufficient quantity.";
      return;
    }

    $inventory->updateOne(
      ['_id' => new MongoDB\BSON\ObjectId($itemForUpdate['_id'])],
      ['$inc' => ['qty' => -$item['qty']]],
      ['session' => $session]
    );

    $subTotal = $item['qty'] * $itemForUpdate['price'];
    $total += $subTotal;
  }

  $receipt = [
    'date' => new DateTime(),
    'items' => $order,
    'total' => $total,
  ];
  
  $orders->insertOne($receipt, ['session' => $session]);

  echo "Order successfully completed and recorded!\nReceipt:\n";
  print_r($receipt);

  $session->commitTransaction();
} catch (MongoDB\Driver\Exception\Exception $e) {
  $session->abortTransaction();
} finally {
  $session->endSession();
}
?>
